examples/polypharmacy_gripnet/model.py

```diff
@@ -1,86 +1,131 @@
+        in_channels = self.encoder.out_channels
+        supergraph = self.encoder.supergraph
+        task_supervertex_name = supergraph.topological_order[-1]
+        num_edge_type = supergraph.supervertex_dict[task_supervertex_name].num_edge_type
+
+        # get the number of nodes on the task-associated supervertex
+        self.num_task_nodes = supergraph.supervertex_dict[task_supervertex_name].num_node
+
+        return MultiRelaInnerProductDecoder(in_channels, num_edge_type)
+
+    def forward(self, edge_index: torch.Tensor, edge_type: torch.Tensor, edge_type_range: torch.Tensor) -> Tuple:
+        x = self.encoder()
+
+        pos_score = self.decoder(x, edge_index, edge_type)
+        pos_loss = -torch.log(pos_score + EPS).mean()
+
+        edge_index = typed_negative_sampling(edge_index, self.num_task_nodes, edge_type_range)
+
+        neg_score = self.decoder(x, edge_index, edge_type)
+        neg_loss = -torch.log(1 - neg_score + EPS).mean()
+
+        loss = pos_loss + neg_loss
+
+        # compute averaged metric scores over edge types
+        num_edge_type = edge_type_range.shape[0]
+        record = []
+        for i in range(num_edge_type):
+            start, end = edge_type_range[i]
+            ps, ns = pos_score[start:end], neg_score[start:end]
+
+            score = torch.cat([ps, ns])
+            target = torch.cat([torch.ones(ps.shape[0]), torch.zeros(ns.shape[0])])
+
+            record.append(list(auprc_auroc_ap(target, score)))
+        auprc, auroc, ap = np.array(record).mean(axis=0)
+
+        return loss, auprc, auroc, ap
+
+    def configure_optimizers(self):
+        optimizer = torch.optim.Adam(self.parameters(), lr=self.conf_solver.BASE_LR)
+
+        return optimizer
+
+    def __step__(self, batch, mode="train"):
+        edge_index, edge_type, edge_type_range = batch
+        loss, auprc, auroc, ap = self.forward(
+            edge_index.reshape((2, -1)), edge_type.flatten(), edge_type_range.reshape((-1, 2))
         )
-        self.dmt = MultiRelaInnerProductDecoder(sum(drug_channels_list), num_drug_edge_relations)
-
-    def forward(
-        self,
-        gene_x,
-        gene_edge_index,
-        gene_edge_weight,
-        gd_edge_index,
-        drug_index,
-        drug_edge_types,
-        drug_edge_range,
-        device,
-    ):
-        z = self.gn(
-            gene_x, gene_edge_index, gene_edge_weight, gd_edge_index, drug_index, drug_edge_types, drug_edge_range
+
+        if mode == "train":
+            self.log(f"{mode}_loss", loss)
+        else:
+            self.log(f"{mode}_auprc", auprc)
+            self.log(f"{mode}_auroc", auroc)
+            self.log(f"{mode}_ap@50", ap)
+
+        return loss
+
+    def training_step(self, batch, batch_idx):
+        return self.__step__(batch)
+
+    def test_step(self, batch, batch_idx):
+        return self.__step__(batch, mode="test")
+
+    # uncomment this function if a valid set is used
```

Comment:
Try not to code like this. Make it a bool option.

Reply:
I deleted this block, as we did not use validation sets in GripNet experiments.

Reply:
Should we keep this as an option in case users of GripNet want a validation set?

Reply:
Yes, it would be better to keep the validation option. In this case, we need to reconstruct the datasets to have a validation set. For this version, I will keep the valid step function, and validate the model on the training set instead.

Reply:
Now line 157 of decode under `GripNetLinkPrediction`? We need to talk this through to ensure I can understand.

Reply:
Clarify a bit in ReadMe